include/boost/geometry/algorithms/detail/overlay/get_turn_info_la.hpp

```diff
@@ -706,35 +663,40 @@ struct get_turn_info_linear_areal
     
     template <bool EnableFirst,
               bool EnableLast,
-              typename Point1,
-              typename Point2,
               typename TurnInfo,
               typename IntersectionInfo,
+              typename UniqueSubRange1,
+              typename UniqueSubRange2,
               typename OutputIterator>
     static inline bool get_turn_info_for_endpoint(
-                            Point1 const& pi, Point1 const& /*pj*/, Point1 const& /*pk*/,
-                            Point2 const& qi, Point2 const& /*qj*/, Point2 const& /*qk*/,
-                            bool is_p_first, bool is_p_last,
-                            bool /*is_q_first*/, bool is_q_last,
+                            UniqueSubRange1 const& range_p,
+                            UniqueSubRange2 const& range_q,
                             TurnInfo const& tp_model,
                             IntersectionInfo const& inters,
                             method_type /*method*/,
                             OutputIterator out)
     {
         namespace ov = overlay;
-        typedef ov::get_turn_info_for_endpoint<AssignPolicy, EnableFirst, EnableLast> get_info_e;
+        typedef ov::get_turn_info_for_endpoint<EnableFirst, EnableLast> get_info_e;
 
         const std::size_t ip_count = inters.i_info().count;
         // no intersection points
-        if ( ip_count == 0 )
+        if (ip_count == 0)
+        {
             return false;
+        }
 
-        if ( !is_p_first && !is_p_last )
+        if (! range_p.is_first() && range_p.size() > 2u)
+        {
+            // P sub-range has no end-points
             return false;
+        }
 
-// TODO: is_q_last could probably be replaced by false and removed from parameters
-
-        linear_intersections intersections(pi, qi, inters.result(), is_p_last, is_q_last);
+        linear_intersections intersections(range_p.at(0),
+                                           range_q.at(0),
+                                           inters.result(),
+                                           range_p.size() == 2u,
+                                           range_q.size() == 2u);
```

Comment:
This is non-intuitive. A person reading it 6 months from now will ask "why 2 exactly"? ;)